### PR TITLE
Config: Add ability to override config variables with env variables

### LINKED ISSUE #25258: Not possible to override backend plugin configuration settings using environment variables
**What happened**:
Tried to use Grafana backend plugin configuration settings using environment variable.

**What you expected to happen**:
Use Grafana backend plugin configuration settings using environment variable should be sent to backend plugin.

**How to reproduce it (as minimally and precisely as possible)**:
For example trying to use `GF_PLUGIN_SOME_PLUGIN_KEY=value` without Grafana configuration file containing it, e.g.
```ini
[plugin.some-plugin]
key = 
```

**More:**
There's functionality for dynamically adding override of env variables for a section using https://github.com/grafana/grafana/blob/01ecbae2ee4898e316429d688035254838eccb49/pkg/setting/setting.go#L1181.

**Environment**:
- Grafana version: 7.0.1 / master

### PR DESCRIPTION
**Pull request by dsotirakis**
**What this PR does / why we need it**:

The purpose of this PR is to be able to override config variables in `default.ini` with environmental variables that follow the pattern:

```
GF_<SectionName>_<KeyName>
```

**Which issue(s) this PR fixes**:

Fixes #25258

**Special notes for your reviewer**:

I tried to use @marefr suggestion using https://github.com/grafana/grafana/blob/01ecbae2ee4898e316429d688035254838eccb49/pkg/setting/setting.go#L1181

but I think that's for assigning variables on the fly?

### CODE REVIEW

File: pkg/plugins/backendplugin/manager/plugin_settings.go [hunk 1]
```diff
@@ -20,6 +24,13 @@ func (ps pluginSettings) ToEnv(prefix string, hostEnv []string) []string {
 	return env
 }
 
+func checkForOverrides(key string, v string) string {
+	if value := os.Getenv(key); value != "" {
+		v = value
+	}
```
Comment (dsotirakis) [inline pkg/plugins/backendplugin/manager/plugin_settings.go:30]: Maybe would be good to have a logger here to print that certain config variables are overriden?

File: pkg/plugins/backendplugin/manager/plugin_settings.go [hunk 2]
```diff
@@ -10,8 +11,13 @@ import (
...
 func (ps pluginSettings) ToEnv(prefix string, hostEnv []string) []string {
-	env := []string{}
+	var env []string
 	for k, v := range ps {
+		key := fmt.Sprintf("%s_%s", prefix, strings.ToUpper(k))
+		if value := os.Getenv(key); value != "" {
+			v = value
+		}
+
 		env = append(env, fmt.Sprintf("%s_%s=%s", prefix, strings.ToUpper(k), v))
```
Comment (aknuds1) [inline pkg/plugins/backendplugin/manager/plugin_settings.go:21]: You can also change this to the following for more readable code:

```suggestion
		env = append(env, fmt.Sprintf("%s=%s", key, v))
```

**Review by marefr — APPROVED** [review]
LGTM with Arve's suggestion applied

**Review by aknuds1 — APPROVED** [review]

File: pkg/plugins/backendplugin/manager/plugin_settings_test.go [hunk 3]
```diff
@@ -41,5 +42,55 @@ func TestPluginSettings(t *testing.T) {
 			require.Len(t, env, 3)
 			require.EqualValues(t, []string{"GF_PLUGIN_KEY1=value1", "GF_PLUGIN_KEY2=value2", "GF_VERSION=6.7.0"}, env)
 		})
+		t.Run("Should override config variable with environmental variable ", func(t *testing.T) {
```
Comment (aknuds1) [inline pkg/plugins/backendplugin/manager/plugin_settings_test.go:45]: Nit, but better to have blank lines between `t.Run` invocations for readability.
```suggestion

		t.Run("Should override config variable with environmental variable ", func(t *testing.T) {
```

File: pkg/plugins/backendplugin/manager/plugin_settings_test.go [hunk 4]
```diff
@@ -41,5 +42,55 @@ func TestPluginSettings(t *testing.T) {
...
+			_ = os.Setenv("GF_PLUGIN_KEY1", "sth")
+			ps := getPluginSettings("plugin", cfg)
+			env := ps.ToEnv("GF_PLUGIN", []string{"GF_VERSION=6.7.0"})
+			sort.Strings(env)
+			require.Len(t, env, 3)
+			require.EqualValues(t, []string{"GF_PLUGIN_KEY1=sth", "GF_PLUGIN_KEY2=value2", "GF_VERSION=6.7.0"}, env)
+
+			defer func() {
```
Comment (aknuds1) [inline pkg/plugins/backendplugin/manager/plugin_settings_test.go:53]: Can you use `t.Cleanup` instead and do it straight after setting the environment variables? When you call `defer` here, it won't execute if the `require`s fail.

File: pkg/plugins/backendplugin/manager/plugin_settings_test.go [hunk 5]
```diff
@@ -41,5 +42,55 @@ func TestPluginSettings(t *testing.T) {
...
+		})
+		t.Run("Config variable doesn't match env variable ", func(t *testing.T) {
+			_ = os.Setenv("GF_PLUGIN_KEY3", "value3")
+			ps := getPluginSettings("plugin", cfg)
+			env := ps.ToEnv("GF_PLUGIN", []string{"GF_VERSION=6.7.0"})
+			sort.Strings(env)
+			require.Len(t, env, 3)
+			require.EqualValues(t, []string{"GF_PLUGIN_KEY1=value1", "GF_PLUGIN_KEY2=value2", "GF_VERSION=6.7.0"}, env)
+
+			defer func() {
```
Comment (aknuds1) [inline pkg/plugins/backendplugin/manager/plugin_settings_test.go:66]: Same as before, use `t.Cleanup` etc.

**Review by aknuds1 — CHANGES_REQUESTED** [review]

File: pkg/plugins/backendplugin/manager/plugin_settings_test.go [hunk 6]
```diff
@@ -41,5 +42,55 @@ func TestPluginSettings(t *testing.T) {
...
+				err := os.Unsetenv("GF_PLUGIN_KEY3")
+				require.NoError(t, err)
+			}()
+		})
+		t.Run("Should override missing config variable with environmental variable ", func(t *testing.T) {
```
Comment (aknuds1) [inline pkg/plugins/backendplugin/manager/plugin_settings_test.go:71]:
```suggestion
		t.Run("Should override missing config variable with environment variable ", func(t *testing.T) {
```

**Review by aknuds1 — CHANGES_REQUESTED** [review]

**Review by aknuds1 — APPROVED** [review]

**Pull request merged**